worker/uniter/runner/context/context.go

```diff
@@ -985,116 +989,86 @@ func (ctx *HookContext) Prepare() error {
+	if ctx.cacheDirty {
+		b = b.UpdateUnitState(ctx.cacheValues)
 	}
 
-	for id, rctx := range ctx.relations {
-		if writeChanges {
-			if e := rctx.WriteSettings(); e != nil {
-				e = errors.Errorf(
-					"could not write settings from %q to relation %d: %v",
-					process, id, e,
-				)
-				logger.Errorf("%v", e)
-				if ctxErr == nil {
-					ctxErr = e
-				}
-			}
+	for _, rctx := range ctx.relations {
+		unitSettings, appSettings := rctx.FinalSettings()
+		if len(unitSettings)+len(appSettings) == 0 {
+			continue // no settings need updating
 		}
+		b = b.UpdateRelationUnitSettings(rctx.RelationTag().String(), unitSettings, appSettings)
 	}
 
-	for rangeKey, rangeInfo := range ctx.pendingPorts {
-		if writeChanges {
-			var e error
-			var op string
-			if rangeInfo.ShouldOpen {
-				e = ctx.unit.OpenPorts(
-					rangeKey.Ports.Protocol,
-					rangeKey.Ports.FromPort,
-					rangeKey.Ports.ToPort,
-				)
-				op = "open"
-			} else {
-				e = ctx.unit.ClosePorts(
-					rangeKey.Ports.Protocol,
-					rangeKey.Ports.FromPort,
-					rangeKey.Ports.ToPort,
-				)
-				op = "close"
-			}
-			if e != nil {
-				e = errors.Annotatef(e, "cannot %s %v", op, rangeKey.Ports)
-				logger.Errorf("%v", e)
-				if ctxErr == nil {
-					ctxErr = e
-				}
-			}
+	for portRange, info := range ctx.pendingPorts {
+		if info.ShouldOpen {
+			b = b.OpenPortRange(portRange.Ports.Protocol, portRange.Ports.FromPort, portRange.Ports.ToPort)
+			continue
 		}
+		b = b.ClosePortRange(portRange.Ports.Protocol, portRange.Ports.FromPort, portRange.Ports.ToPort)
 	}
 
-	// add storage to unit dynamically
-	if len(ctx.storageAddConstraints) > 0 && writeChanges {
-		err := ctx.unit.AddStorage(ctx.storageAddConstraints)
-		if err != nil {
-			err = errors.Annotatef(err, "cannot add storage")
+	// Generate change request but skip its execution if no changes are pending.
+	consolidateReq, numChanges := b.Build()
+	if numChanges > 0 {
+		if err := ctx.unit.ConsolidateHookChanges(consolidateReq); err != nil {
+			err = errors.Annotatef(err, "cannot apply changes")
 			logger.Errorf("%v", err)
-			if ctxErr == nil {
-				ctxErr = err
-			}
+			return err
 		}
 	}
 
-	if ctx.podSpecYaml != nil && writeChanges {
-		err := ctx.commitPodSpec()
-		if ctxErr == nil {
-			ctxErr = err
+	// Call completed successfully; update local state
+	ctx.cacheDirty = false
+
+	// add storage to unit dynamically
+	if len(ctx.storageAddConstraints) > 0 {
+		if err := ctx.unit.AddStorage(ctx.storageAddConstraints); err != nil {
+			err = errors.Annotatef(err, "cannot add storage")
+			logger.Errorf("%v", err)
+			return err
 		}
 	}
 
-	// TODO (tasdomas) 2014 09 03: context finalization needs to modified to apply all
-	//                             changes in one api call to minimize the risk
-	//                             of partial failures.
-
-	if !writeChanges {
-		return ctxErr
+	if ctx.podSpecYaml != nil {
+		return ctx.commitPodSpec()
 	}
 
-	return ctxErr
+	return nil
```

Comment:
So these are things that are still outside the 'commit-in-one-transaction' right?

Reply:
Yes, they were kept out to make this PR easier to review. I am working on a followup PR to move these into the batch call.